### DISCUSSION TITLE
Jupyter: Specify path to Kernel

### DISCUSSION THREAD

**Discussion by rigelifland** [microsoft/vscode-jupyter#10404] [Questions and Answers]
I'd like to be able to directly specify the path to a kernel for Jupyter, rather than only have the option to select it from a list. Preferably this would be something that can be set in the workspace preferences or the *.ipynb itself. It doesn't seem like this is possible with VSCode currently - is this an external technical limitation that is imposed by Jupyter, or should I make a feature request?

Background:
I'm managing a large number of projects (500+) each of which have their own virtualenvs. I'd like to be able to open VSCode inside one of these projects and use Jupyter Notebooks without having to install every single kernel and search through a giant list of the venvs. This search is made especially difficult since the naming convention for all the virtual environments is to name them all ".venv". If was was able to specify the path to the Kernel in the workspace settings or the ipynb file it would solve my problem.

**Comment by DonJayamanne** [comment]
I'm trying to better understand your workflow, please could you clarify my understanding of the problem:
* You have folders `project1/.venv`, `project2/.venv`, `project2/.venv`
* You open the folder `project2` in VS Code
* You open a notebook and you will get a list of kernels `.venv` from this folder.
	* However you also have other virtual environments `folder2/.venv` in the list of kernels is that correct?
	* This shouldn't be the case.

If i understand this correctly, you are some how ending up with the kernels `folder1/.venv` `folder2/.venv` in the kernel dropdown list and thats way too many, and would like to only see the kernel thats specific to that folder. Is that right.

> I'd like to be able to open VSCode inside one of these projects and use Jupyter Notebooks without having to install every single kernel and search through a giant list of the venvs.

Assume you opened the project `folder2` do you want to use the kernle in .venv sub folder or would you like to use a single kernel for all of these 500 folders?
If you want to use the same kernel for all of these 500 folders, you can then :
* Install a kernel in the global kernels location using the command `python -m ipykernel install --user --name myGlobalKernel --display-name myGlobalKernel` and in the kernel picker this will appear & you can easily search for this by typing in `myGlobalKernel` or you could use some other name.

**Reply by rigelifland** [reply]
> you are some how ending up with the kernels folder1/.venv folder2/.venv in the kernel dropdown list and thats way too many, and would like to only see the kernel thats specific to that folder.

This is correct. Expanding on the example setup with Project1 and Project2, in order for the kernels to show up in a notebook at all, they have to be installed. However if you run `python -m ipykernel install <options>` it always gets installed in a common folder:

> _https://ipython.readthedocs.io/en/stable/install/kernel_install.html_
system-wide (e.g. /usr/local/share),
in Jupyter’s env (sys.prefix/share),
per-user (~/.local/share or ~/Library/share)

This means that if I install the kernel from project1/.venv, regardless of where I open VSCode that kernel will show up. Since I need separate venvs (and ipykernel installs) for each project, it leads to a very large list of kernels to look through. 

Just to be super clear, I'm talking about the kernel in Jupyter Notebooks:
<img width="279" alt="Screen Shot 2022-06-09 at 9 11 57 PM" src="https://user-images.githubusercontent.com/21069069/172988858-d18dff70-b42d-4ddd-9770-76f178d240c0.png">

Not the workspace python interpreter:
<img width="404" alt="Screen Shot 2022-06-09 at 9 12 48 PM" src="https://user-images.githubusercontent.com/21069069/172988928-cf46c7ef-d785-4312-bc97-36de9e40899c.png">

The behavior I'm looking for is actually modeled by VSCode's automatic detection and selection of python interpreters. If you have a venv in the workspace it will automatically detect it and add it to the list,  if not use it by default outright. If I go to another workspace, it no longer shows in the list.

Having thought about it more, perhaps this is actually a limitation of ipykernel's install methods? Since it can only be installed to common folders, what I'm asking for is a means of using the ipykernel without installing it (to be clear, it will be `pip install`'d, but not `python -m ipykernel install`'d.) Hope this makes sense.

**Comment by GMNGeoffrey** [comment]
Was this ever resolved? I'm experiencing a similar issue. I have several folders open, each with their own venv called ".venv". With a new notebook I see the following options when selecting a kernel:

![image](https://github.com/microsoft/vscode-jupyter/assets/5732088/0944756e-0bf2-405a-8480-85434ea3a72c)

That's obviously not very helpful. It sounds from your comment @DonJayamanne that this is not a state I should be getting into at all, although I still think that the kernel picker should disambiguate with absolute paths (or perhaps always show them when a kernel is selected). VSCode generally does a good job of that in other places.

Apologies for necroing this old discussion. I came here on the direction of the bug report form at https://github.com/microsoft/vscode-jupyter/issues/new/choose

**Reply by DonJayamanne** [reply]
@GMNGeoffrey please can you file a new issue.

**Reply by GMNGeoffrey** [reply]
https://github.com/microsoft/vscode-jupyter/issues/15066

**Comment by DonJayamanne** [comment]
Closing this discussion as resolved
See here for solution https://github.com/microsoft/vscode/issues/130946#issuecomment-1899389049
